src/lem/circuit.rs

```diff
@@ -0,0 +1,902 @@
+    hash2: PathTicker,
+    hash3: PathTicker,
+    hash4: PathTicker,
+}
+
+impl SlotsCounter {
+    #[inline]
+    pub(crate) fn next_hash2(&mut self, path: Path) -> usize {
+        self.hash2.tick(path)
+    }
+
+    #[inline]
+    pub(crate) fn next_hash3(&mut self, path: Path) -> usize {
+        self.hash3.tick(path)
+    }
+
+    #[inline]
+    pub(crate) fn next_hash4(&mut self, path: Path) -> usize {
+        self.hash4.tick(path)
+    }
+
+    pub(crate) fn cont(&mut self, new: Path, from: &Path) {
+        self.hash2.cont(new.clone(), from);
+        self.hash3.cont(new.clone(), from);
+        self.hash4.cont(new, from);
+    }
+}
+
+/// Contains a `slots_map` that maps `LEMOP`s to their slots. This map is not
+/// expected to be injective, as two or more `LEMOP`s can be mapped to the same
+/// slot.
+///
+/// The `slots` attribute is derived from `slots_map` by iterating on its values.
+pub struct SlotsInfo {
+    slots_map: IndexMap<LEMOP, (usize, SlotType)>,
+    slots: IndexSet<(usize, SlotType)>,
+}
+
+impl SlotsInfo {
+    #[inline]
+    pub fn get_slot(&self, op: &LEMOP) -> Result<&(usize, SlotType)> {
+        self.slots_map
+            .get(op)
+            .ok_or_else(|| anyhow!("Slot not found for LEMOP {:?}", op))
+    }
+}
+
+impl LEMOP {
+    /// STEP 1: compute the slot mapping on a first (and unique) traversal
+    ///
+    /// While traversing parallel paths, we need to reuse compatible slots. For
+    /// example, on a `MatchTag` with two arms, such that each arm is a `Hash2`,
+    /// we want those hashes to occupy the same slot. We do this by using a
+    /// `SlotsCounter`, which can return the next index for a slot category
+    /// (`Hash2`, `Hash3` etc).
+    ///
+    /// Further, when we:
+    /// * Construct the same pointer twice
+    /// * Destruct the same pointer twice
+    /// * Construct a pointer that was previously destructed
+    /// * Destruct a pointer that was previously constructed
+    ///
+    /// We want to reuse the same slot as before. To accomplish this, we use
+    /// hashmaps that can recover the slots that were previously allocated.
+    pub fn slots_info(&self) -> Result<SlotsInfo> {
+        let mut slots_map = IndexMap::default();
+        let mut slots = IndexSet::default();
+
+        // these hashmaps keep track of slots that were allocated for preimages
+        let mut preimgs2_map: HashMap<&[MetaPtr; 2], (usize, SlotType)> = HashMap::default();
+        let mut preimgs3_map: HashMap<&[MetaPtr; 3], (usize, SlotType)> = HashMap::default();
+        let mut preimgs4_map: HashMap<&[MetaPtr; 4], (usize, SlotType)> = HashMap::default();
+
+        // these hashmaps keep track of slots that were allocated for images
+        let mut imgs2_map: HashMap<&MetaPtr, (usize, SlotType)> = HashMap::default();
+        let mut imgs3_map: HashMap<&MetaPtr, (usize, SlotType)> = HashMap::default();
+        let mut imgs4_map: HashMap<&MetaPtr, (usize, SlotType)> = HashMap::default();
```

Comment:
Hmm, I'm not sure that's a perf optimization. The first version of the code was as you described. But then I decided to add the `SlotType` to avoid calls to `SlotType::from_lemop` (which does a `match` on `LEMOP`) and also to make the code cleaner after retrieving data from those hashmaps

Reply:
You actually don’t need the `from_lemop` if it’s expensive: just add a slot type to the input, aligned with the other arguments:
```
( $preimg: expr, $img: expr, $preimgs_map: expr, $imgs_map: expr, $counter_fn: expr, $slot_type ) => {
                    match ($preimgs_map.get($preimg), $imgs_map.get($img)) {
                        (Some(u), _) | (_, Some(u)) => {
                            // reusing a slot
                            if slots_map.insert(op.clone(), (*u, $slot_type)).is_some() {
                                bail!("Duplicated LEMOP: {:?}", op)
                            }
                        }
```
But again, I do recognize what you have now is a bit more intuitive, and so fine to keep until the perf is needed.